### ISSUE TITLE
CMSSignedData.replaceSigners() does not handle DigestAlgorithms parameters properly

### ISSUE THREAD

**Issue by bsanchezb** [bcgit/bc-java#1585]
Hello,

I'm currently working on implementation of CAdES-E-ERS ([ETSI TS 119 122-3](https://www.etsi.org/deliver/etsi_ts/119100_119199/11912203/01.01.01_60/ts_11912203v010101p.pdf)) for my project, which requires computation of CMS digest to be protected by an embedded evidence-record or its validation.

As an evidence-record is placed within _UnsignedAttributes_, I need to remove the attribute in order to compute the original hash of the covered CMS, which involves re-computation of the CMS. What I'm currently doing is I remove the evidence-record attribute from _UnsignedAttributes_ of a corresponding _SignerInformation_, re-compute collection of all present _SignerInformations_ in a form of _SignerInformationStore_, and call **CMSSignedData.replaceSigners(originalCMS, newSignerInformationStore)**, so it looks like:
```
CMSSignedData originalSignedData = ...
List<SignerInformation> newSignerInformationList = new ArrayList<>();
for (SignerInformation signerInformation : cmsSignedData.getSignerInfos().getSigners()) {
    AttributeTable unsignedAttributes = signerInformation.getUnsignedAttributes();
    if (unsignedAttributesContainER(unsignedAttributes)) {
        unsignedAttributes = unsignedAttributes.remove(erAttributeIdentifier);
        signerInformation = SignerInformation.replaceUnsignedAttributes(signerInformation, unsignedAttributes);
    }
    newSignerInformationList.add(signerInformation);
}
CMSSignedData newSignedData = CMSSignedData.replaceSigners(originalSignedData , new SignerInformationStore(newSignerInformationList));
```

The method [CMSSignedData.replaceSigners(cmsSignedData, signerInformationStore)](https://github.com/bcgit/bc-java/blob/9279d29588e52d8f0260c2ce5e730c773c247569/pkix/src/main/java/org/bouncycastle/cms/CMSSignedData.java#L591) seems to cause a problem on _digestAlgs_ set computation, ignoring algorithm parameters. I.e. when processing a CMS having _DigestAlgorithms_ defined like:
![image](https://github.com/bcgit/bc-java/assets/23361149/c6894d4f-0986-4ba3-a343-f629009bed1e)
the call of _CMSSignedData.replaceSigners_ method produces a signature with the parameters attribute being removed:
![image](https://github.com/bcgit/bc-java/assets/23361149/606bfad7-7f99-4db6-963b-d96bacec1fab)

This of course **modifies the binary content of the signature**, which impacts the hash computation, making it impossible to verify the embedded evidence-record.

The issue can be originally caused by [CMSUtils.addDigestAlgs()](https://github.com/bcgit/bc-java/blob/main/pkix/src/main/java/org/bouncycastle/cms/CMSUtils.java#L155) method, which ignores (?) the embedded parameters.

Worth noticing, that the _SignerInformation_ in question has a _DigestAlgorithm_ attribute defined in the same way, before and after its replacement within the CMS:
![image](https://github.com/bcgit/bc-java/assets/23361149/5e622bd7-cb4e-461b-8b22-aec4fe04e49a)

Thus, I find the _SignedData.digestAlgorithms_ field to be built incorrectly.

Kind regards,
Aleksandr

**Comment by bsanchezb** [comment]
> Strictly speaking the parameters should be absent. I agree this is not a really feature though. I'll have to look into this - the DigestAlgorithms is supposed to represent a set of all the digest algorithms in use, the re-encoding is happening as the set is being rebuilt.

Indeed, I understand that there is no interest in having NULL parameters, but as we do not have control over signatures provided to our app, we should be able to handle all of them. In fact, all solutions implementing the embedded ERs that we currently found, have this problem. Therefore, it is important to support these cases.

> I know this will sound a bit crazy, but I'd assume if you call CMSSignedData.replaceSigners() twice, the first time with the original signer information set, things actually work out?

Maybe I did not understand the suggestion, but I was still not able to produce a new CMS with NULL parameters within _DigestAlgorithms_ field. I tried to call _CMSSignedData.replaceSigners()_ with the original (not modified) set of _SignerInformations_ and simply execute it twice in order.

**Comment by dghgit** [comment]
This should now be fixed on https://www.bouncycastle.org/betas

**Comment by bsanchezb** [comment]
Hi @dghgit , I confirm the issue has been fixed.

Thank you.

**Comment by dghgit** [comment]
Fixed in 1.79.
